src/api/callback.cc

```diff
@@ -158,20 +158,33 @@ MaybeLocal<Value> InternalMakeCallback(Environment* env,
     CHECK(!argv[i].IsEmpty());
 #endif
 
-  InternalCallbackScope scope(env, resource, asyncContext);
+  Local<Function> hook_cb = env->async_hooks_callback_trampoline();
+  int flags = InternalCallbackScope::kNoFlags;
+  int hook_count = 0;
+  if (!hook_cb.IsEmpty()) {
+    flags = InternalCallbackScope::kSkipAsyncHooks;
+    AsyncHooks* async_hooks = env->async_hooks();
+    hook_count = async_hooks->fields()[AsyncHooks::kBefore] +
+                 async_hooks->fields()[AsyncHooks::kAfter];
+  }
+
+  InternalCallbackScope scope(env, resource, asyncContext, flags);
   if (scope.Failed()) {
     return MaybeLocal<Value>();
   }
 
-  Local<Function> domain_cb = env->domain_callback();
   MaybeLocal<Value> ret;
-  if (asyncContext.async_id != 0 || domain_cb.IsEmpty()) {
-    ret = callback->Call(env->context(), recv, argc, argv);
+
+  if (hook_count != 0) {
+    MaybeStackBuffer<Local<Value>, 16> args(2 + argc);
+    args[0] = v8::Number::New(env->isolate(), asyncContext.async_id);
+    args[1] = callback;
+    for (int i = 0; i < argc; i++) {
+      args[i + 2] = argv[i];
+    }
+    ret = hook_cb->Call(env->context(), recv, args.length(), &args[0]);
   } else {
-    std::vector<Local<Value>> args(1 + argc);
-    args[0] = callback;
-    std::copy(&argv[0], &argv[argc], args.begin() + 1);
-    ret = domain_cb->Call(env->context(), recv, args.size(), &args[0]);
+    ret = callback->Call(env->context(), recv, argc, argv);
```

Comment:
I took me a while to understand why the domain callback is not skipped in this case - simply because domains use an async hook with before/after set. If they would move to `executionAsyncResource` and only installing an `init` hook like `AsyncLocalStorage` this would be broken. I assume this will not happen as domains are deprecated.
Not sure if we should add some note about this. Or is it safe to assume that there are tests in place to ensure this want break?

Reply:
I chose to go for the least possible logical change to avoid significant changes to domains. While they _are_ considered deprecated, there _are_ users in the wild and I don't want to break them especially given how little we _trust_ our current domain tests. I _would_ like to redo domains later to use `AsyncLocalStorage`, but before that I hope to have a more proper way to do callback wrapping that can be generalized for different hook needs.